packages/ckeditor5-html-support/src/emptyblocks.ts

```diff
@@ -0,0 +1,100 @@
+/**
+ * @license Copyright (c) 2003-2025, CKSource Holding sp. z o.o. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-licensing-options
+ */
+
+/**
+ * @module html-support/emptyblocks
+ */
+
+import { Plugin } from 'ckeditor5/src/core.js';
+import type { UpcastElementEvent, Element } from 'ckeditor5/src/engine.js';
+
+const EMPTY_BLOCK_MODEL_ATTRIBUTE = 'htmlEmptyBlock';
+
+/**
+ * This plugin allows for preserving empty block elements in the editor content instead of
+ * automatically filling them with block fillers (`&nbsp;`).
+ *
+ * Empty elements are detected during upcast and marked with a special attribute.
+ * During downcast, elements with this attribute have their `getFillerOffset` set to `null`
+ * which prevents adding block fillers.
+ *
+ * This is useful when you want to:
+ *
+ *	* Preserve empty block elements exactly as they were in the source HTML
+ * 	* Allow for styling empty blocks with CSS (block fillers can interfere with height/margin)
+ * 	* Maintain compatibility with external systems that expect empty blocks to remain empty
+ *
+ * For example, this allows for HTML like:
+ *
+ * ```html
+ * <p></p>
+ * <p class="spacer"></p>
+ * <td></td>
+ * ```
+ * to remain empty instead of being converted to:
+ *
+ * ```html
+ * <p>&nbsp;</p>
+ * <p class="spacer">&nbsp;</p>
+ * <td>&nbsp;</td>
+ * ```
+ */
+export default class EmptyBlocks extends Plugin {
+	/**
+	 * @inheritDoc
+	 */
+	public static get pluginName() {
+		return 'EmptyBlocks' as const;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public static override get isOfficialPlugin(): true {
+		return true;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public init(): void {
+		const editor = this.editor;
+		const schema = editor.model.schema;
+
+		// Register the attribute for block elements
+		schema.extend( '$block', {
+			allowAttributes: [ EMPTY_BLOCK_MODEL_ATTRIBUTE ]
+		} );
```

Comment:
Have you considered extending the `$container`? What about table cells? Those are not `$block`.